lib/data/file/backend.js

```diff
@@ -65,11 +74,15 @@ export const backend = {
                 return callback(errors.InternalError);
             })
             .on('end', () => {
-                log.debug('finished writing data', { key });
-                fs.fsync(fd, () => {
-                    fs.close(fd);
-                    return callback(null, key);
-                });
+                streamEnd = true;
+                // Only in flowing mode when fs.write finished
+                if (request._readableState.flowing) {
```

Comment:
I don't think we should be relying on the internals of the readable stream to know if our last write is finished.  Instead, why don't we have a variable writeFinished initially set to false and it will only be set to true in the fs.write callback.  So when end is called on the readable stream, we only close the file and call the callback if the writeFinished is set to true.  If it is false, we keep checking back after some interval until it is finished and then close and call the callback.  @GiorgioRegni , does that make sense to you?